host/windows/syscall.c

```diff
@@ -312,9 +462,87 @@ int oe_syscall_rmdir_ocall(const char* pathname)
 **==============================================================================
 */
 
+#define OE_SOCKET_FD_MAGIC 0x29b4a345c7564b57
+typedef struct win_socket_fd
+{
+    uint64_t magic;
+    SOCKET socket;
+} oe_socket_fd_t;
+
+static oe_socket_fd_t _invalid_socket = {OE_SOCKET_FD_MAGIC, INVALID_SOCKET};
+
+oe_host_fd_t _make_socket_fd(SOCKET sock)
+{
+    oe_host_fd_t fd = (oe_host_fd_t)&_invalid_socket;
+    if (sock != INVALID_SOCKET)
+    {
+        oe_socket_fd_t* socket_fd =
+            (oe_socket_fd_t*)malloc(sizeof(oe_socket_fd_t));
+        socket_fd->magic = OE_SOCKET_FD_MAGIC;
+        socket_fd->socket = sock;
+        fd = (oe_host_fd_t)socket_fd;
+    }
+    return fd;
+}
+
+SOCKET _get_socket(oe_host_fd_t fd)
+{
+    oe_socket_fd_t* socket_fd = (oe_socket_fd_t*)fd;
+    if (socket_fd && socket_fd->magic == OE_SOCKET_FD_MAGIC)
+        return socket_fd->socket;
+    return INVALID_SOCKET;
+}
+
+static oe_host_fd_t _dup_socket(oe_host_fd_t oldfd)
```

Comment:
Is there a better way to clone a socket in Windows? I had put this in as a placeholder.

Reply:
It sort of depends on what the expectation is. There is a function WSADuplicateSocket but it is designed for sharing sockets across processes. But that isn't the purpose of dup in Linux (I'm not really sure I understand the purpose of dup actually).

I'll do some more research.

Reply:
https://stackoverflow.com/questions/11635219/dup2-dup-why-would-i-need-to-duplicate-a-file-descriptor

Reply:
Used WSADuplicateSocket with current process. As I understand it, the goal of dup is that if one fd is closed, the duplicate is not closed.

I need to make sure WSADuplicateSocket actually creates a reference to the same socket, rather than just making a separate socket with the same parameters.

Reply:
I don't know if it helps, but this was our implementation back in Apache Mesos: https://github.com/apache/mesos/blob/5c50403eb233991eaa8f83b5bc50eae2c33d5850/3rdparty/stout/include/stout/os/windows/dup.hpp#L25

Reply:
Ok cool, that is also what I did.

I found this in Win32 docs that confirms it does what I'm expecing:

> All of the state information associated with a socket is held in common across all the descriptors because the socket descriptors are duplicated and not the actual socket.